File: usaco/section_2.4/cowtour.cpp
/*
ID: huangch7
LANG: C++
TASK: cowtour
 */


#include <iostream>
#include <vector>
#include <map>
#include <algorithm>
#include <cstring>
#include <string>
#include<sstream>
#include<cstdio>
#include <cmath>
#include <iomanip> 

using namespace std;

const string task_name = "cowtour";

  // 计算两点之间的距离
double get_dist(int i, int j, vector<pair<int, int> > pos) {
    int dx = pos[i].first - pos[j].first;
    int dy = pos[i].second - pos[j].second;
    return sqrt(dx * dx + dy * dy);
  };


int main() {
  freopen("cowtour.in","r",stdin);
  freopen("cowtour.out","w",stdout);

  // 读取输入数据
  int n;
  cin >> n;
  vector<pair<int, int> > pos(n);
  vector<string > conn;
  
  for (int i = 0; i < n; i++) {
    cin >> pos[i].first >> pos[i].second;
  }
    string line;
    std::getline(std::cin, line);
  
  for (int i = 0; i < n; i++) {
    //    cin >> line;
    std::getline(std::cin, line);

    conn.push_back(line);
  }
  //  for (int i = 0; i<conn.size(); ++i){
  //    cout << conn[i] <<endl;
  //  }


  // 计算所有点之间的最短距离
  vector<vector<double> > dist(n, vector<double>(n, -1));
  for (int i = 0; i < n; i++) {
    for (int j = 0; j < n; j++) {
      //      cout << "debug " << i <<"," <<j<< " " << conn[i] << " " << conn[i][j] <<endl;
      if (conn[i][j] == '0') {
        continue;
      }
      dist[i][j] = get_dist(i, j, pos);
    }
    dist[i][i] = 0;
  }
//  for (int i = 0; i < n; i ++){
//    for (int j = 0; j < n; j ++) {
//        cout << dist[i][j] << " ";
//    }
//    cout << endl;
//  }

  // 使用 Floyd-Warshall 算法计算所有点之间最短距离
  for (int k = 0; k < n; k++) {
    for (int i = 0; i < n; i++) {
      for (int j = 0; j < n; j++) {
        if (dist[i][k] < -0.5 || dist[k][j] < -0.5 ) {
          continue;
        }
        if (dist[i][j] <-0.5 || dist[i][j] > dist[i][k] + dist[k][j]) {
          dist[i][j] = dist[i][k] + dist[k][j];
        }
      }
    }
  }

//  for (int i = 0; i < n; i ++){
//    for (int j = 0; j < n; j ++) {
//        cout << dist[i][j] << " ";
//    }
//    cout << endl;
//  }

  // 计算两个不同颜色点的最大距离
  double ans = -1;
  vector<int> colors(n, -1);
  
  int cnt = 0;
  for (int i = 0; i < n; i++) {
    if (colors[i] > -0.5) {
      continue;
    }
    colors[i] = cnt++;
    for (int j = 0; j < n; j++) {
      if (dist[i][j] > -0.5) {
        colors[j] = colors[i];
      }
    }
  }
  // 计算每个点的最大距离
  vector<double> max_dist(n, 0);
  vector<double> max_field_dist(n, 0);
  for (int i = 0; i < n; i++) {
    for (int j = 0; j < n; j++) {
      if (dist[i][j] < -0.5) {
        continue;
      }
      if (max_dist[i] < dist[i][j]) {
        max_dist[i] = dist[i][j];
      }
    }
    if (max_dist[i] > max_field_dist[colors[i]])
      max_field_dist[colors[i]] = max_dist[i];
    
  }
  
//  for (int i = 0; i< n; ++){
//  cout<< colors[i] << endl;
//    }
  for (int i = 0; i < n; i++) {
    for (int j = 0; j < n; j++) {
      if (colors[i] == colors[j]) {
        continue;
      }
      double cur = max_dist[i] + max_dist[j] + get_dist(i, j, pos);
      cur = max(cur, max_field_dist[colors[i]]);
      cur = max(cur, max_field_dist[colors[j]]);
      if (ans < -0.5 || ans > cur) {
        ans = cur;
      }
    }
  }

  // 输出答案
  //cout << setprecision(6) << ans << endl;
  printf("%.6f\n", ans);

  return 0;
}

File: usaco/section_2.4/cowtour.py
"""
ID: huangch7
LANG: PYTHON3
TASK: cowtour
"""
import math
task_name = "cowtour"

fin = open(f'{task_name}.in', 'r')
fout = open(f'{task_name}.out', 'w')

pos = []
conn = []
n = int(fin.readline())
dist = [[None] * n for _ in range(n)]

for _ in range(n):
    pos.append([int(x) for x in fin.readline().strip().split()])
for _ in range(n):
    conn.append(fin.readline())

def get_dist(i, j):
    x = pos[i][0] - pos[j][0]
    y = pos[i][1] - pos[j][1]
    return math.sqrt(x * x + y * y)

for i in range(n):
    for j in range(n):
        if conn[i][j] == '0':
            continue

        dist[i][j] = get_dist(i, j)
    dist[i][i] = 0
                              

for k in range(n):
    for i in range(n):
        for j in range(n):
            
            if dist[i][k] is not None and dist[k][j] is not None:
                if dist[i][j] is None or dist[i][j] > dist[i][k] + dist[k][j]:
                    dist[i][j] = dist[i][k] + dist[k][j]

# max dist from x
max_dist = {}
for i in range(n):
    best = 0
    for j in range(n):
        if dist[i][j] is None:
            continue
        if best < dist[i][j]:
            best = dist[i][j]
    max_dist[i] = best

# count parts
cnt = 0
colors = {}
for i in range(n):
    if i in colors:
        continue
    colors[i] = cnt
    cnt += 1
    for j in range(n):
        if dist[i][j] is not None:
            colors[j] = colors[i]
# print(colors)
# for line in dist:
#     print(line)

ans = None
for i in range(n):
    for j in range(n):
        if colors[i] == colors[j]:
            continue
        cur = max_dist[i] + max_dist[j] + get_dist(i, j)
        if ans is None or ans > cur:
            ans = cur
fout.write(f"{ans:.6f}\n")
fout.close()
